src/entity_ref.rs

```diff
@@ -294,3 +349,83 @@ impl<'a, T: Component> ComponentRef<'a> for &'a mut T {
 pub trait ComponentRefShared<'a>: ComponentRef<'a> {}
 
 impl<'a, T: Component> ComponentRefShared<'a> for &'a T {}
+
+struct ComponentBorrow<'a> {
+    archetype: &'a Archetype,
+    /// State index for the borrowed component in the `archetype`.
+    state: usize,
+}
+
+impl<'a> ComponentBorrow<'a> {
+    // This method is unsafe as the `index` parameter is not validated
+    // to actually point to the correct component in the `archetype`.
```

Comment:
This is a bit confusing. No in-bounds value of `index` is "incorrect", and an out of bounds value does not address an incorrect *component*, but rather causes immediate undefined behavior. Bounds checking is what's being omitted here; we always have the right component type due to `Archetype::get_state::<T>`.

Reply:
Addressed again!

I hope that the `{ComponentBorrow, ComponentBorrowMut}::for_component()` documentation is okay. I would have cribbed the documentation for `{Ref, RefMut}::new()` if there was any. If it's not okay, would you be able to suggest a comment I could copy paste for them? I'm not super familiar with how this part works or the safety guarantees.